src/osp/Active/SysSceneGraph.h

```diff
@@ -0,0 +1,188 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2020 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "basic.h"
+#include "activetypes.h"
+
+#include <algorithm>
+
+namespace osp::active
+{
+
+/**
+ * @brief Helps add entities to a reserved space in ACtxSceneGraph
+ *
+ * All reserved space must be used, or else this class will assert on
+ * destruction.
+ */
+class SubtreeBuilder
+{
+public:
+    constexpr SubtreeBuilder(ACtxSceneGraph& rScnGraph, ActiveEnt root, TreePos_t first, TreePos_t last)
+     : m_rScnGraph{rScnGraph}
+     , m_root{root}
+     , m_first{first}
+     , m_last{last}
+    { }
+    constexpr SubtreeBuilder(SubtreeBuilder&& move) = default;
+    SubtreeBuilder(SubtreeBuilder const& copy) = delete;
+    ~SubtreeBuilder() { assert(m_first == m_last); }
+
+    /**
+     * @brief Add a child that might have more descendants
+     *
+     * @return
+     */
+    [[nodiscard]] SubtreeBuilder add_child(ActiveEnt ent, uint32_t descendantCount);
+
+    void add_child(ActiveEnt ent)
+    {
+        (void) add_child(ent, 0);
+    }
+
+    std::size_t remaining()
+    {
+        return m_last - m_first;
```

Comment:
minor: assert that first is not larger than last. It's a class invarient, as far as I can tell.